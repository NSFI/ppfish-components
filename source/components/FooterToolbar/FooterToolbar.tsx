import * as React from 'react';
import classNames from 'classnames';
import { addEventListener } from '../../utils';

interface FooterToolbarProps {
  children: React.ReactNode;
  prefixCls?: string;
  className?: string;
  target?: () => HTMLElement | Window | null;
  style?: React.CSSProperties;
}

const FooterToolbar: React.FC<FooterToolbarProps> = props => {
  const { children, className, prefixCls, style, target } = props;
  const [offset, setOffset] = React.useState(0);

  const wrapperRef = React.useRef(null);

  const setToolbarPosition = () => {
    const wrapper = wrapperRef.current;
    const { height } = getComputedStyle(wrapper);
    const wrapperHeight = Number.parseInt(height);

    let targetElement: null | Window | HTMLElement = document.documentElement;
    if (target && typeof target === 'function' && target() !== window) {
      targetElement = target();
    }
    const offsetObj = {
      containerHeight: (targetElement as HTMLElement).clientHeight,
      containerScrollTop: (targetElement as HTMLElement).scrollTop,
    };

    const offset = offsetObj.containerHeight + offsetObj.containerScrollTop - wrapperHeight;
    const maxOffset = (targetElement as HTMLElement).scrollHeight - wrapperHeight;
    const nextOffset = offset > maxOffset ? maxOffset : offset;
    setOffset(nextOffset);
  };

  React.useEffect(() => {
    setToolbarPosition();
    const target = props.target?.() || window;
    const scrollListener = addEventListener(target, 'scroll', () => {
      setToolbarPosition();
    });

    const resizeListener = addEventListener(window, 'resize', () => {
      setToolbarPosition();
    });

    return () => {
      scrollListener?.remove?.();

      resizeListener?.remove?.();
    };
  }, []);

  const toolbarStyle: React.CSSProperties = {
    ...style,
    position: 'absolute',
    top: offset,
  };
  return (
    <div
      className={classNames({ [className]: true, [prefixCls]: true })}
      style={toolbarStyle}
      ref={wrapperRef}
    >
      <div className={`${prefixCls}-inner`}>{children}</div>
    </div>
  );
};

FooterToolbar.defaultProps = {
  children: null,
  className: '',
  prefixCls: 'fishd-footer-toolbar',
};

export default FooterToolbar;
