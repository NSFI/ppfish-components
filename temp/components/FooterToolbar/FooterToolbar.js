var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { addEventListener } from '../../utils';
var FooterToolbar = /** @class */ (function (_super) {
    __extends(FooterToolbar, _super);
    function FooterToolbar(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            offset: 0
        };
        _this.setToolbarPosition = function () {
            var target = _this.props.target;
            var height = getComputedStyle(_this.wrapper).height;
            var wrapperHeight = Number.parseInt(height);
            var targetElement = document.documentElement;
            if (target && typeof target === 'function' && target() !== window) {
                targetElement = target();
            }
            var offsetObj = {
                containerHeight: targetElement.clientHeight,
                containerScrollTop: targetElement.scrollTop
            };
            var offset = offsetObj.containerHeight + offsetObj.containerScrollTop - wrapperHeight;
            var maxOffset = targetElement.scrollHeight - wrapperHeight;
            _this.setState({
                offset: offset > maxOffset ? maxOffset : offset
            });
        };
        return _this;
    }
    FooterToolbar.prototype.componentDidMount = function () {
        var _this = this;
        this.setToolbarPosition();
        var target = (this.props.target && this.props.target()) || window;
        this.scrollListener = addEventListener(target, 'scroll', function () {
            _this.setToolbarPosition();
        });
        this.resizeListener = addEventListener(window, 'resize', function () {
            _this.setToolbarPosition();
        });
    };
    FooterToolbar.prototype.componentWillUnmount = function () {
        if (this.scrollListener) {
            this.scrollListener.remove();
        }
        if (this.resizeListener) {
            this.resizeListener.remove();
        }
    };
    FooterToolbar.prototype.render = function () {
        var _a;
        var _this = this;
        var _b = this.props, children = _b.children, className = _b.className, prefixCls = _b.prefixCls, style = _b.style;
        var toolbarStyle = __assign(__assign({}, style), { position: 'absolute', top: this.state.offset });
        return (React.createElement("div", { className: classNames((_a = {}, _a[className] = true, _a[prefixCls] = true, _a)), style: toolbarStyle, ref: function (wrapper) {
                _this.wrapper = wrapper;
            } },
            React.createElement("div", { className: prefixCls + "-inner" }, children)));
    };
    FooterToolbar.defaultProps = {
        children: null,
        className: '',
        prefixCls: 'fishd-footer-toolbar'
    };
    FooterToolbar.propTypes = {
        children: PropTypes.node,
        className: PropTypes.string,
        prefixCls: PropTypes.string,
        target: PropTypes.func,
        style: PropTypes.object
    };
    return FooterToolbar;
}(React.Component));
export default FooterToolbar;
