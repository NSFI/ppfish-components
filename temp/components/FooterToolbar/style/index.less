@import (reference) "../../../assets/css/index";

@fishd-toolbar-prefix: ~"@{fishd-prefix}-footer-toolbar";
.@{fishd-toolbar-prefix} {
  width: 100%;
  background: @component-background;
  border-top: 1px solid @border-color-split;
  &-inner {
    margin: 0 auto;
    padding: 10px 20px;
    min-height: 52px;
    width: 100%;
  }
}
