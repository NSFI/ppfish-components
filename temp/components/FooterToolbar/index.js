import FooterToolbar from './FooterToolbar';
import './style/index.less';
export default FooterToolbar;
