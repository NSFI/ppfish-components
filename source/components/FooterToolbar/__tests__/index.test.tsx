import React from 'react';
import { mount } from 'enzyme';
import FooterToolbar from '../FooterToolbar';

describe('FooterToolbar', () => {
  let originGetComputedStyle;
  beforeAll(() => {
    originGetComputedStyle = Object.getOwnPropertyDescriptor(window, 'getComputedStyle').value;
    Object.defineProperty(window, 'getComputedStyle', {
      value: () => ({
        height: 100,
      }),
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: originGetComputedStyle,
    });
  });

  it('FooterToolbar should render', () => {
    const Page = () => {
      return (
        <div style={{ position: 'relative' }}>
          <FooterToolbar>footer</FooterToolbar>
        </div>
      );
    };
    const wrapper = mount(<Page />);
    expect(wrapper).toMatchSnapshot();
  });
});
