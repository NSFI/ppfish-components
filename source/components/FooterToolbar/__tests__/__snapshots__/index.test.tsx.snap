// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterToolbar FooterToolbar should render 1`] = `
<Page>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <FooterToolbar
      className=""
      prefixCls="fishd-footer-toolbar"
    >
      <div
        className=" fishd-footer-toolbar"
        style={
          Object {
            "position": "absolute",
            "top": -100,
          }
        }
      >
        <div
          className="fishd-footer-toolbar-inner"
        >
          footer
        </div>
      </div>
    </FooterToolbar>
  </div>
</Page>
`;
